cli/lint_serialize.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+//! This module provides a utility to wrap LintDiagnostic
+//! into a JSON serializable struct.
+use deno_lint::diagnostic::LintDiagnostic;
+use serde::Serialize;
+
+#[derive(Serialize)]
+struct SerializableLocation {
```

Comment:
**What's going on here?**

Since the `LintDiagnostic` struct does not implement the `Serialize` trait, **and** it's defined in a separate crate, we need to use adapters to make create a "serializable" version of it. These `SerializableX` structs are simply adapters to the `deno_lint` structs, that also `derive(Serialize)`.

Reply:
Makes sense. Because `deno_lint` is part of Deno (https://github.com/denoland/deno_lint), we could just make the `deno_lint::diagnostic::LintDiagnostic`  serialisable upstream so we don't need this wrapper. Would save a minuscule amount of memory and CPU and would simplify this a little. @bartlomieju WDYT?

Reply:
`LintDiagnostic` should already be JSON-serializable, if it's not, then we should make it so in `deno_lint`

Reply:
> `LintDiagnostic` should already be JSON-serializable

Yeah, unfortunately it doesn't look like this is the case. 😞 

```
error[E0277]: the trait bound `deno_lint::diagnostic::LintDiagnostic: serde::Serialize` is not satisfied
   --> cli/lint.rs:242:3
    |
242 |   lints: Vec<LintDiagnostic>,
    |   ^^^^^ the trait `serde::Serialize` is not implemented for `deno_lint::diagnostic::LintDiagnostic`\
```

Also `deno_lint` doesn't have `serde` as a dependency, so I'm not sure how eager we are to add one just for this :shrug:

Having `deno_lint::LintDiagnostic` serializable and using that here would be the simplest, but it's important to be aware of the tradeoff. If we made `LintDiagnostic` serializable and just used that, then a change to the struct in `LintDiagnostic` would change the API of `deno lint`. Thankfully we have an integration test covering the expected API, but that test is in a separate project, meaning that it would be fairly easy to merge changes to `deno_lint` without realizing they would "break" things (hopefully just pipelines).

- Are we sure we like the simplicity over control tradeoff here?
- Is this something we want to take care of now or iteratively in a follow-up?

Completely your call what you want to do here. Just let me know :smile:

Reply:
> would save a minuscule amount of memory

@lucacasonato FWIW, I'm not entirely sure we'd save memory because we'd still need to clone the `LintDiagnostic` before pushing it to the `Vec` 🤔

I'm not super familiar with managing the life cycle of these things though. Please let me know if I'm missing something 😅

Reply:
@souldzin it's okay to do that in `deno_lint`. I've opened an issue: https://github.com/denoland/deno_lint/issues/253 Will try to get it done soon

Reply:
> I've opened an issue: denoland/deno_lint#253 Will try to get it done soon

Thanks @bartlomieju! Sounds good to me :smile:

If you are not able to get to it, I might be able to jump on it as well. Just let me know :smile:

Reply:
@souldzin if you want to work on it, then go ahead, I described details in the issue.

Reply:
Thanks @bartlomieju for the upstream fixed. Rebased with `master` and removed the `Serializable` wrappers 👍